content/doc/developer/cli/writing-cli-commands.adoc

```diff
@@ -0,0 +1,65 @@
+layout: developersection
+---
+
+Plugins can contribute additional commands to link:/doc/book/managing/cli/[Jenkins CLI] which.
+
+This is useful for:
+
+* exposing administrative commands to admins, so that they can script some of the Jenkins babysitting work, 
+* exposing data and operations to builds executing inside Jenkins, so that they can interact with Jenkins in a richer way.
+
+Writing commands can be done in two ways.
+
+## Annotating a method with CLIMethod
+
+If you are exposing behaviors of your model objects as CLI commands, the easiest way to achieve it is to put link:http://javadoc.jenkins-ci.org/byShortName/CLIMethod[@CLIMethod] on a method of your model object. See link:https://javadoc.jenkins-ci.org/hudson/model/Queue.html#clear--[Queue.clear()] as an example. 
+In addition to the command name as specified in the annotation, you also need to define "CLI.command-name.shortDescription" as a message resource, which captures one line human-readable explanation of the command (see link:https://javadoc.jenkins.io/hudson/cli/CLICommand.html#getShortDescription--[CLICommand.getShortDescription()]).
+
+[source,java]
+----
+public class AbstractItem {
+    @CLIMethod(name="delete-job")
+    public synchronized void delete() throws IOException, InterruptedException {
+        performDelete();
+ 
+        if (this instanceof TopLevelItem) {
+            Jenkins.get().deleteJob((TopLevelItem)this);
+        }
+        Jenkins.get().rebuildDependencyGraph();
+    }
+    ...
+}
+----
+
+Notice that the method is an instance method. So when the delete-job command is executed, which job is deleted? 
+To resolve this, you also need to define a `CLI resolver`, which uses a portion of arguments and options to determine the instance object that receives a method call.
+
+[source,java]
+----
+@CLIResolver
+public static AbstractItem resolveForCLI(
+        @Argument(required=true,metaVar="NAME",usage="Job name") String name) throws CmdLineException {
+    AbstractItem item = Jenkins.get().getItemByFullName(name, AbstractItem.class);
+    if (item==null)
+        throw new CmdLineException(null,"No such job exists:"+name);
+    return item;
+}
+----
+
+Of all the resolver methods that are discovered, Jenkins picks the one that returns the best return type. 
+It doesn't matter where the resolver method is defined, or how it's named.
+
+Both resolver methods and CLI methods can have any number of link:https://args4j.kohsuke.org[args4j] annotations, which causes the parameters and arguments to be injected upon a method invocation. 
+All the other unannotated parameters receive null. 
+Combined with the stapler method binding, this enables you to make your method invokable from both CLI and HTTP.
+
+## Extending CLICommand
+
+You can also implement a CLI commmand as a subtype of link:http://javadoc.jenkins-ci.org/byShortName/CLICommand[CLICommand], and put link:http://javadoc.jenkins-ci.org/byShortName/Extension[@Extension]. 
```

Comment:
Links are dead `byShortName` looks to be defunct

Reply:
Some work, e.g. `CLIMethod` (the other one on this line).

Reply:
Certainly not a problem for this content migration, as e.g. `jenkinsdoc` has dedicated support for this URL type.

Reply:
I had to qualify `Extension` to make it work, everything else seemed to work for core, `Channel` is in the remoting component so I guess it doesn't work for that